### PR TITLE
#733: ENG-3337: Add Jepsen counter test

### PR DESCRIPTION
**Pull request by ttyusupov**
https://github.com/YugaByte/yugabyte-db/issues/733

Used https://github.com/riptano/jepsen/blob/cassandra/cassandra/src/cassandra/counter.clj as a base for counter test.

Taken `checker.counter` changes from https://github.com/riptano/jepsen/blob/master/jepsen/src/jepsen/checker.clj#L182 and updated to be compatible with Jepsen framework 0.3.3.

### CODE REVIEW

File: jepsen/project.clj [hunk 1]
```diff
@@ -9,7 +9,7 @@
                  [org.clojure/tools.cli "0.3.5"]
                  [clj-time "0.11.0"]
                  [jepsen.txn "0.1.0"]
-                 [knossos "0.3.3-yb" :exclusions [org.slf4j/slf4j-log4j12]]
+                 [knossos "0.3.3-yb-SNAPSHOT" :exclusions [org.slf4j/slf4j-log4j12]]
```
Comment (amitanandaiyer) [inline jepsen/project.clj]: Would probably be better if we don't depend on a SNAPSHOT build. Is it possible to bump up and push the next version, so we don't acccidentally break this if we were to push code to knossos?

File: jepsen/src/jepsen/checker.clj [hunk 2]
```diff
@@ -359,20 +359,18 @@
 (defn counter
   "A counter starts at zero; add operations should increment it by that much,
   and reads should return the present value. This checker validates that at
-  each read, the value is at greater than the sum of all :ok increments, and
-  lower than the sum of all attempted increments.
+  each read, the value is at greater than the sum of all :ok increments and
+  :invoke decrements, and lower than the sum of all attempted increments and
```
Comment (amitanandaiyer) [inline jepsen/src/jepsen/checker.clj]: use :invoke/attempted for both increments/decrements. Unless they are meant to mean different things.

File: jepsen/src/jepsen/checker.clj [hunk 3]
```diff
@@ -382,34 +380,51 @@
              upper              0             ; Upper bound on counter value
              pending-reads      {}            ; Process ID -> [lower read-val]
```
Comment (amitanandaiyer) [inline jepsen/src/jepsen/checker.clj]: Does the comment need to be updated? It looks like we are storing [lower, upper] instead.

Comment (amitanandaiyer) [reply jepsen/src/jepsen/checker.clj]: on line 395.

File: jepsen/src/jepsen/checker.clj [hunk 4]
```diff
@@ -382,34 +380,51 @@
...
+              [:invoke :read]
+              (recur history lower upper
+                     (assoc pending-reads (:process op) [[lower upper]])
+                     reads)
+
+              [:ok :read]
+              (let [read-ranges (get pending-reads (:process op))
+                    v (:value op)
+                    [l' u'] (first read-ranges)
+                    read (or (some (fn [[l u]] (when (<= l v u) [l v u])) read-ranges)
```
Comment (amitanandaiyer) [inline jepsen/src/jepsen/checker.clj]: I'm slightly confused here. Can you pls explain what you are doing?

read is assigned to the first matching [l u] which satisfies (<= l v u).
Why do that? Doesn't (:process op) tell us the op that the value is for?

Comment (ttyusupov) [reply jepsen/src/jepsen/checker.clj]: I've added more comments to clarify how this checker works. As for `(:process op)` - this is ID of the process, not the value. We are trying to find a range within possible ranges which contains the value `v` we've received from read op. If we can't find it - this is an inconsistency and we just save first range and `v` into `reads`, so later when we check for errors we see that `l <= v <= r` is not satisfied and report an error.

File: jepsen/src/jepsen/checker.clj [hunk 5]
```diff
@@ -359,57 +359,88 @@
...
-
-                (recur history lower upper pending-reads reads))))))))
+                       (dissoc pending-reads process-id)
+                       (conj reads read)))
+
+              [:invoke :add]
+              (let [value (:value op)
+                    [l' u'] (if (> value 0) [lower (+ upper value)] [(+ lower value) upper])]
+                (recur history l' u'
+                       (reduce-kv #(assoc %1 %2 (conj %3 [l' u'])) {} pending-reads)
```
Comment (amitanandaiyer) [inline jepsen/src/jepsen/checker.clj]: I don't understand this.

I was expected something like 
(reduce-kv #(assoc %1 %2 [ (min (first %3) l') (max (second %3) u')])) {} pending-reads)

So, we are updating for each k, v (= [l u]) in pending-writes as  k -> v = [min (l, ,') max (u u')].

Comment (ttyusupov) [reply jepsen/src/jepsen/checker.clj]: It simply adds `[l' u']` range to list of ranges for each process in `pending-reads`:
```
jepsen.cli=> (let [
        #_=> upper 7 value -1 lower 3
        #_=> pending-reads {:1 [[1 2]] :2 [[2 4]]}
        #_=> [l' u'] (if (> value 0) [lower (+ upper value)] [(+ lower value) upper])]
        #_=> (reduce-kv #(assoc %1 %2 (conj %3 [l' u'])) {} pending-reads))
{:1 [[1 2] [2 7]], :2 [[2 4] [2 7]]}
```

I think this implementation which I simply ported from Cassandra's test suite could be improved, but for this iteration, it should be good enough.

**Pull request merged**